### ISSUE TITLE
Question regarding the dag-cbor encoding of objects (trying to reproduce in Python)

### ISSUE THREAD

**Issue by AndreaCensi** [ipld/ipld#65]
Hi, 

I am trying to re-implement the CID encoding in Python and I am having trouble reproducing the results given by `ipfs dag put`. 

My understanding is that `ipfs dag put` does the following steps (assuming default choices for hash and format):
1. Start with a cbor-encodable object ob (read by default in JSON format)
2. Get the CBOR representation.
3. Compute the SHA256 of those bytes.
4. Create a multihash (code=18).
5. Create a CID, with parameters version = `1`, encoding = `dag-cbor`, and the multihash obtained above. 

My translation of the above as Python code is as follows.
Here I use the libraries `multihash` and `py-cid`.

```python
def get_cbor_dag_hash(ob):
    """ Returns a base58 string of a CID pointing to ob expressed in CBOR format"""
    import cbor2
    import hashlib
    import multihash
    from cid.cid import make_cid
    ob_cbor = cbor2.dumps(ob)
    ob_cbor_hash = hashlib.sha256(ob_cbor).digest()
    mh = multihash.encode(digest=ob_cbor_hash, code=18)
    cid = make_cid(1, 'dag-cbor', mh)
    return cid.encode().decode('ascii')
```

With this code I get the expected result for the object `{}` (empty dictionary), which codes to `zdpuAyTBnYSugBZhqJuLsNpzjmAjSmxDqBbtAqXMtsvxiN2v3`.

However I get some divergence on some other test vectors:

```
object = 42
ipfs dag put -> zdpuAvVUB18y6k8jZTZUHSq4FctExZm4Mxkt3xEyYW8ptQqru
my code -> zdpuAu1CyNiDwRXh6uSwRYdAHZjuhrC7pbQi4jvQHuNTPWxAe
```

```
object = {"a": 1}
ipfs dag put -> zdpuAxTWBh3d49ZCgPP44BT8AAa9qiqxB167yZCFdxHxVg6gN
my code -> zdpuB2H7FsgxU1PVuGcwk4TDYGgv7xrQ7naFJ1YHfRV71t7Rf
```

I am particularly puzzled because my code agrees with `ipfs dag put` on one instance - so I know that most of the code is right, and probably I only miss some detail about how objects are encoded.

 Could you provide some hint on what I misunderstand about the format?

thanks!

**Comment by Stebalien** [comment]
@warpfork does refmt canonicalize the number types?

**Comment by AndreaCensi** [comment]
@Stebalien Yes, I think it has to do with number types. I tried with objects, strings, arrays, and null, and it works fine with those types.

**Comment by vmx** [comment]
@AndreaCensi For all sorts of fun with numbers, please also see https://github.com/ipld/interface-ipld-format/issues/9#issuecomment-431029329 and https://github.com/ipld/specs/issues/80.

**Comment by Stebalien** [comment]
Yeah, it looks like python is using the canonical encoding and we aren't. We really need to make a decision here.

**Comment by warpfork** [comment]
Probably it would be good to look at the raw bytes being fed to the hashing, as the final CID is quite opaque.

There's not a whole lot of canonicalization to _do_ with basic integers.  The byte for encoding the integer 1 is.... "0x01".

**Comment by AndreaCensi** [comment]
@warpfork I found the explanation. It has to do with `ipfs dag put` converting integers to floats internally.

In fact, I verified that `ipfs dag put` encodes `0` and `0.0` to the same hash `zdpuB39X55PgAai6zTPtTGtSrArCZh5ow1zE1d8ZGM6wwhPrZ`, and `1` and `1.0` to the same hash, and so on. It seems like it is converting all integers to float internally. 

Instead, the Python implementation above is type-safe, as it encodes `int` to CBOR integers and `float` to CBOR floats. 

I understand that the distinction between int and float is not well defined in JSON, but in CBOR, number types are well defined. So in the spirit of IPLD being a foundational technology, I would suggest to rely on the CBOR data types. Especially numbers. 

I recently had to take a close look at the JSON messiness, and so I will also mention that CBOR has `bytes`, while JSON doesn't have it. When people need to move around bytes, they do their own encoding (e.g. base64) on top of JSON strings, which has a performance hit in addition to not being elegant. 

CBOR in general is pretty good (as it came after JSON and people knew what problems to fix...).

**Comment by Stebalien** [comment]
Oh. Damn. This has nothing to do with CBOR and everything to do with JSON. The problem is that we're encoding the data as _json_, shipping it off to go-ipfs, decoding it as json (converting to floats in the process), and then encoding those to CBOR (correctly as floats).

Your code is correct. The following will produce the same output:

```go
package main

import (
	"fmt"

	cbor "github.com/ipfs/go-ipld-cbor"
)

func main() {
	obj1, err := cbor.WrapObject(42, ^uint64(0), -1)
	if err != nil {
		panic(err)
	}
	fmt.Println(obj1.Cid())

	obj2, err := cbor.WrapObject(map[string]int{"a": 1}, ^uint64(0), -1)
	if err != nil {
		panic(err)
	}
	fmt.Println(obj2.Cid())
}
```
